blog/2022/04/simple-chains.md

```diff
@@ -0,0 +1,489 @@
+ 44.241192 seconds (369.06 k allocations: 62.864 GiB, 2.55% gc time)
+
+julia> eval_loss_accuracy(train_loader, model, device),
+       eval_loss_accuracy(test_loader, model, device)
+((loss = 0.072f0, acc = 97.7883), (loss = 0.0642f0, acc = 97.99))
+```
+
+#### How long did SimpleChains.jl take?
+
+SimpleChains on an AMD EPYC 7513 32-Core Processor:
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 34.410432 seconds (55.84 M allocations: 5.920 GiB, 3.79% gc time, 85.95% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.972, 0.093898475f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9744, 0.08624289f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  3.083624 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9835666666666667, 0.056287352f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9831, 0.053463124f0)
+```
+
+SimpleChains on an Intel i9 10980XE 18-Core Processor:
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 34.915917 seconds (83.99 M allocations: 5.431 GiB, 3.91% gc time, 95.22% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.972, 0.09354472f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.975, 0.084895514f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  1.324148 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9836833333333334, 0.054299146f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9829, 0.053510584f0)
+```
+
+SimpleChains on an Intel i7 1165G7 4-Core Processor (thin and light laptop CPU):
+
+```julia
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+ 42.664702 seconds (103.71 M allocations: 5.236 GiB, 2.85% gc time, 77.60% compilation time)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.9863166666666666, 0.044742946f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9841, 0.048268355f0)
+
+julia> @time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+  6.320631 seconds
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+(0.99485, 0.017716348f0)
+
+julia> SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+(0.9888, 0.035394628f0)
+```
+
+Note that smaller batch sizes improve accuracy per epoch, and batch sizes were set to be proportional to the number of threads.
+
+## Benchmark Summary
+
+Latency before the first epoch begins training is problematic, but SimpleChains.jl is fast once compiled.
+Post-compilation, the 10980XE was around 70% faster than Flux with the A100 GPU.
```

Comment:
```suggestion
Post-compilation, the 10980XE was competitive with Flux using an A100 GPU, and about 35% faster than the V100.
```
I realized the loaders weren't putting the data onto the GPU up front. Doing this improved the Flux GPU times (in the suggestions above), so my advantage is smaller.

But that makes Flux look quite good compared to PyTorch on the GPU, which was oddly slow.
This also suggests that PyTorch's problem isn't that the data wasn't already on the GPU, as Flux was many times faster when experiencing that same problem.